File: spring/14852.py
n = int(input())


if n == 1:
	print(2)
elif n == 2:
	print(7)
else:
	arr = []
	arr.append(2)
	arr.append(7)
	tmp = 1
	for i in range(2, n):
		arr.append((arr[i-1] * 2 + arr[i-2] * 3 + 2 * tmp) % 1000000007)
		tmp += arr[i-2]
	print(arr[-1])

File: spring/14852.cpp
#include <iostream>

int main(void) {
    int n;
    scanf("%d", &n);

    long long int arr[n];

    if (n == 1)
        printf("2\n");
    else if (n==2)
        printf("7\n");
    else {
        arr[0] = 2, arr[1] = 7;
        long long int tmp = 1;
        for (int i=2; i<n; i++) {
            arr[i] = (arr[i-1] * 2 + arr[i-2] * 3 + 2 * tmp) % 1000000007;
            tmp += arr[i-2];
        }
        printf("%lld\n", arr[n-1]);
    }
}